### PR TITLE
don't UB on dangling ptr deref, instead check inbounds on projections

### PR DESCRIPTION
**Pull request by RalfJung**
This implements https://github.com/rust-lang/reference/pull/1387 in Miri. See that PR for what the change is about.

Detecting dangling references in `let x = &...;` is now done by validity checking only, so some tests need to have validity checking enabled. There is no longer inherently a "nodangle" check in evaluating the expression `&*ptr` (aside from the aliasing model).

r? @oli-obk 

Based on:
- https://github.com/rust-lang/reference/pull/1387
- https://github.com/rust-lang/rust/pull/115524

### CODE REVIEW

File: compiler/rustc_const_eval/messages.ftl [hunk 1]
```diff
@@ -282,7 +281,7 @@ const_eval_pointer_out_of_bounds =
         *[many] bytes
     } starting at offset {$ptr_offset} is out-of-bounds
 const_eval_pointer_use_after_free =
-    pointer to {$allocation} was dereferenced after this allocation got freed
+    {$bad_pointer_message}: {$alloc_id} has been freed, so pointer is dangling
```
Comment (saethlin) [inline compiler/rustc_const_eval/messages.ftl:284]:
```suggestion
    {$bad_pointer_message}: {$alloc_id} has been freed, so this pointer is dangling
```

File: tests/ui/consts/const-eval/dangling.rs [hunk 2]
```diff
@@ -1,10 +0,0 @@
-use std::mem;
-
-// Make sure we error with the right kind of error on a too large slice.
-const TEST: () = { unsafe {
-    let slice: *const [u8] = mem::transmute((1usize, usize::MAX));
-    let _val = &*slice; //~ ERROR: evaluation of constant value failed
-    //~| slice is bigger than largest supported object
-} };
```
Comment (RalfJung) [inline tests/ui/consts/const-eval/dangling.rs:8]: This test no longer errors since `*` no longer checks that the place is dereferenceable (and CTFE doesn't do validity checks that would catch the invalid reference).

File: compiler/rustc_const_eval/messages.ftl [hunk 3]
```diff
@@ -282,7 +281,7 @@ const_eval_pointer_out_of_bounds =
         *[many] bytes
     } starting at offset {$ptr_offset} is out-of-bounds
 const_eval_pointer_use_after_free =
-    pointer to {$allocation} was dereferenced after this allocation got freed
```
Comment (RalfJung) [inline compiler/rustc_const_eval/messages.ftl:285]: Half-way through I realized that this says "dereferenced" even when it is just doing an in-bounds pointer arithmetic check. It was too late to easily split that fix into a separate commit -- sorry.

File: compiler/rustc_const_eval/src/interpret/projection.rs [hunk 4]
```diff
@@ -203,23 +202,25 @@ where
...
-        }))
+        // Due to lifetime issues, we have to buffer the fields in a `Vec`:
+        // We cannot keep borrowing `self` during iteration (users will want to mutate `InterpCx` during iteration),
+        // and yet we need `self` for `offset` (because it needs to do a bounds check).
+        let fields: Vec<_> = (0..len)
+            .map(move |i| {
+                // `Size` multiplication
+                base.offset(stride * i, field_layout, self)
+            })
+            .collect();
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/projection.rs:222]: If this is too bad for perf, we could in principle do a single bounds-check upfront and then use non-bounds-checked arithmetic -- but that would require poking some holes into the `Projectable` abstraction so I'd rather not do that...

**Comment by RalfJung** [comment]
So the stress test actually got faster, but a bunch of other benchmarks get slower. I thought of a way to keep lazy iteration, let's see if that helps. If not, the regression is probably truly caused by doing an inbounds check for each element (rather than a single check upfront), and that will be harder to avoid.

@bors try @rust-timer queue

**Comment by RalfJung** [comment]
That barely seems to have helped, bummer. Looks like we'll need a way to avoid doing the in-bounds check for each element.

File: compiler/rustc_const_eval/src/interpret/projection.rs [hunk 5]
```diff
@@ -37,31 +38,32 @@ pub trait Projectable<'tcx, Prov: Provenance>: Sized + std::fmt::Debug {
...
 
     /// Offset the value by the given amount, replacing the layout and metadata.
-    fn offset_with_meta(
+    fn offset_with_meta<'mir, M: Machine<'mir, 'tcx, Provenance = Prov>>(
         &self,
         offset: Size,
         meta: MemPlaceMeta<Prov>,
         layout: TyAndLayout<'tcx>,
-        cx: &impl HasDataLayout,
+        ecx: &InterpCx<'mir, 'tcx, M>,
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/projection.rs:46]: My plan is to add a `OffsetMode` enum parameter controlling whether we use inbounds or nowrap (or maybe wrapping) offset.

**Comment by RalfJung** [comment]
That helped less than I hoped, but it does seem to have helped.

**Comment by RalfJung** [comment]
This is now updated to the new wording in https://github.com/rust-lang/reference/pull/1387. However it is still blocked.

I wonder what this does with perf...
@bors try @rust-timer queue

**Comment by RalfJung** [comment]
Hm, so that's more tests being affected, and it seems worse for the ctfe test, but no primary test regressions this time... neither option seems clearly better.

**Comment by RalfJung** [comment]
This is not clearly better, and it's more code duplication, so I'll revert https://github.com/rust-lang/rust/pull/114330/commits/6d02d7dd01d493adb7df9c239b72b3714e71d80c again.

**Comment by RalfJung** [comment]
@oli-obk this is finally ready. :)
@rustbot ready

**Review by oli-obk — APPROVED** [review]
r=me with or without test re-added

File: tests/ui/consts/const-eval/dangling.rs (hunk 2 above)
Comment (RalfJung) [reply tests/ui/consts/const-eval/dangling.rs:8]: I can add it back as a `-Zextra-const-ub-checks` test.

Comment (oli-obk) [reply tests/ui/consts/const-eval/dangling.rs:8]: I guess I meant leaving it in as a normal test that now passes, with a note that it could stop passing at any time and that's not a breaking change. But `-Zextra-const-ub-checks` is also good.

Comment (RalfJung) [reply tests/ui/consts/const-eval/dangling.rs:8]: This also checks that the error is not reported without the flag, so we will notice if behavior changes again.

**Comment by RalfJung** [comment]
Some regressions were expected base on prior benchmarks, this is as good as I managed to get it.  Only secondary benchmarks are affected.

**Comment by pnkfelix** [comment]
Visiting for weekly perf triage

* From skimming the PR, one can see that the PR author (RalfJung) iterated on this to identify a solution that would minimize regressions.
* As noted by the PR author, only secondary benchmarks were affected.
* Also, while instruction-counts regressed, the [cycle-counts](https://perf.rust-lang.org/compare.html?start=a00c09e9d80b763fb29206b47b04e1d99c3ace96&end=e7bdc5f9f869219e8d20060b42a09ea10a837851&stat=cycles%3Au)
  did not, at least not enough to pass our noise threshold.
* marking as triaged.

@rustbot label: +perf-regression-triaged
